splunklogger.js

```diff
@@ -451,82 +556,43 @@ SplunkLogger.prototype._sendEvents = function(context, callback) {
  * @throws Will throw an error if the <code>context</code> parameter is malformed.
  * @public
  */
-SplunkLogger.prototype.send = function (context, callback) {
-    callback = callback || function(){};
+SplunkLogger.prototype.send = function(context, callback) {
     context = this._initializeContext(context);
     
+    // Store the context, and its estimated length
     this.contextQueue.push(context);
+    this.eventsBatchSize += Buffer.byteLength(JSON.stringify(this._makeBody(context)), "utf8");
 
-    if (context.config.autoFlush) {
-        this.flush(callback);
-    }
-    else {
-        callback();
+    var batchOverSize = this.eventsBatchSize > this.config.maxBatchSize && this.config.maxBatchSize > 0;
+    var batchOverCount = this.contextQueue.length >= this.config.maxBatchCount && this.config.maxBatchCount > 0;
+
+    // Only flush if not using manual batching, and if the contextQueue is too large or has many events
+    if (this.config.autoFlush && (batchOverSize || batchOverCount)) {
+        this.flush(callback || function(){});
     }
 };
 
 /**
- * Manually send events in <code>this.contextQueue</code> to Splunk, after
- * chaining any <code>functions</code> in <code>this.middlewares</code>.
- * Auto flush settings will be used from <code>this.config.autoFlush</code>,
- * ignoring auto flush settings every on every <code>context</code> in <code>this.contextQueue</code>.
+ * Manually send all events in <code>this.contextQueue</code> to Splunk.
  *
  * @param {function} [callback] - A callback function: <code>function(err, response, body)</code>.
  * @public
  */
-SplunkLogger.prototype.flush = function (callback) {
+SplunkLogger.prototype.flush = function(callback) {
     callback = callback || function(){};
 
     var context = {};
 
-    // Use the batching setting from this.config
-    if (this.config.autoFlush) {
-        // TODO: handle case of multiple events with autoFlush off, flushing fast
-        // Just take the oldest event in the queue
-        context = this.contextQueue.pop();
-    }
-    else {
-        // Empty the event queue
-        var queue = this.contextQueue;
-        this.contextQueue = [];
-        var data = "";
-        for (var i = 0; i < queue.length; i++) {
-            data += JSON.stringify(this._makeBody(queue[i]));
-        }
-        context.message = data;
-    }
-    
-    // Initialize the context, then manually set the data
-    context = this._initializeContext(context);
-    
-    // Copy over the middlewares
-    var callbacks = [];
-    for (var j = 0; j < this.middlewares.length; j++) {
-        callbacks[j] = this.middlewares[j];
+    // Empty the queue, reset the eventsBatchSize
+    var queue = this.contextQueue;
+    this.contextQueue = [];
+    this.eventsBatchSize = 0;
+    var data = "";
+    for (var i = 0; i < queue.length; i++) {
+        data += JSON.stringify(this._makeBody(queue[i]));
```

Comment:
You've already stringified it once when you counted the bytes - why are you doing it again?